presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java

```diff
@@ -42,6 +45,38 @@ public static Expression canonicalizeExpression(Expression expression)
     private static class Visitor
             extends ExpressionRewriter<Void>
     {
+        @Override
+        public Expression rewriteComparisonExpression(ComparisonExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            // if we have a comparison of the form <constant> <op> <expr>, normalize it to
+            // <expr> <op-flipped> <constant>
+            if (node.getLeft() instanceof Literal && !(node.getRight() instanceof Literal)) {
+                return new ComparisonExpression(
+                        node.getOperator().flip(),
+                        treeRewriter.rewrite(node.getRight(), context),
+                        treeRewriter.rewrite(node.getLeft(), context));
+            }
+
+            return treeRewriter.defaultRewrite(node, context);
+        }
+
+        @Override
+        public Expression rewriteArithmeticBinary(ArithmeticBinaryExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (node.getOperator() == ArithmeticBinaryExpression.Operator.MULTIPLY || node.getOperator() == ArithmeticBinaryExpression.Operator.ADD) {
+                // if we have a operation of the form <constant> [+|*] <expr>, normalize it to
+                // <expr> [+|*] <constant>
+                if (node.getLeft() instanceof Literal && !(node.getRight() instanceof Literal)) {
+                    return new ArithmeticBinaryExpression(
+                            node.getOperator(),
+                            treeRewriter.rewrite(node.getRight(), context),
+                            treeRewriter.rewrite(node.getLeft(), context));
+                }
+            }
+
+            return treeRewriter.defaultRewrite(node, context);
+        }
+
```

Comment:
what about `io.prestosql.sql.tree.LogicalBinaryExpression`? With logical operation it would be better to have literal on the left side and so to make it coherent we could have literal on the left side for arithmetic and comparison operation as well.

Reply:
We could go either way. The purpose of this change is to be able to remove duplicate terms in logical expressions, so that would just be for cosmetics.

Reply:
Anyway `literal AND expr` or `literal OR expr` should be optimized during planning with constant folding, so this rewrite does not need to care about it.